WIP: Annotate com.google.common.base package

dilraj45 opened the pull request:
Annotate the com.google.common.base package for nullness type system.

mernst commented on guava/src/com/google/common/base/CaseFormat.java:
```diff
@@ -133,6 +133,9 @@ public final String to(CaseFormat format, String str) {
   @SuppressWarnings({
     "upperbound:assignment.type.incompatible", "upperbound:compound.assignment.type.incompatible", // refinement on multiple assignments in expression
     "upperbound:argument.type.incompatible", // length of sequence with ArrayLenRange in offset
+    "dereference.of.nullable" // Variable 'out' will never be dereferenced when it is null. If block
```
How about moving the initialization of `out` from line 145 to line 139 where it is declared?
Then I think the suppression is not necessary.

mernst commented on guava/src/com/google/common/base/Converter.java:
```diff
@@ -165,12 +167,23 @@ protected Converter() {
-  B correctedDoForward(@Nullable A a) {
+  @SuppressWarnings({
+    "return.type.incompatible",
+    /*
```
Your code is inconsistent about whether it uses /* or // style comments.  I suggest // except for Javadoc.  It's easier to read and more compact.

mernst commented on guava/src/com/google/common/base/Converter.java:
```diff
@@ -165,12 +167,23 @@ protected Converter() {
+    /*
+     * If handleNullAutomatically is true then correctedDoForward returns null only if 'a' is null
+     * else it returns a non-null value. Known behaviour of @PolyNull with ternary operator, please
+     * refer to link: https://github.com/typetools/checker-framework/issues/1170
```
This whole comment can be simplified to:

```
bug in handling of @PolyNull with ternary operator: https://github.com/typetools/checker-framework/issues/1170
```

mernst commented on guava/src/com/google/common/base/Converter.java:
```diff
@@ -165,12 +167,23 @@ protected Converter() {
+     */
+    "argument.type.incompatible"
+    /*
+     * If handleNullAutomatically is false, then argument can be assumed to be non-null
```
Why can this be assumed?  Please explain the reasoning.

dilraj45 replied:
Here is the link to comments on `handleNullAutomatically` in class [com.google.common.base.Converter](https://github.com/google/guava/blob/master/guava/src/com/google/common/base/Converter.java), see, https://github.com/google/guava/blob/master/guava/src/com/google/common/base/Converter.java#L128-L131. If documentation needs refinement, please let me know so that I can pull request the changes for the same.

mernst commented on guava/src/com/google/common/base/Converter.java:
```diff
@@ -386,7 +406,19 @@ public String toString() {
+    "override.return.invalid",
+    "override.param.invalid"
+    /*
+     * This method can return null value even if only type A is declared to be nullable and null is
```
This seems incorrect to me.  A function that returns B should not return null unless B is a nullable type.
Furthermore, if A is a non-null type, then it should not be permitted to pass null into `apply`.

dilraj45 replied:
Yes, this method can return a null value even if B is not declared to be nullable. It depends upon the type of argument that is supplied. I'm not sure how to depict this behaviour using annotations. Should this be reported as a bug?

mernst replied:
Where does the documentation say that the method is allowed to return a different type than B?

dilraj45 replied:
By adding an overriding `@PolyNull` annotation we are describing that this method may return null even if B is declared to be `@NonNull`. It depends entirely on whether a null is passed as an argument to this method or not.  Nullness Checker will expects a nullable value whenever a nullable value is passed to this method (even if B is declared to be `@NonNull`). This is conflicting with the overridden method, which expects it to be annotated as ```public final B apply (A a)```. But if we do so then we will be missing on the case when B is declared as `@NonNull` and A as `@Nullable` and this method still might return a nullable value. Is this the way we should be handling this? Please let me know your thoughts on this

mernst commented on guava/src/com/google/common/base/Converter.java:
```diff
@@ -423,30 +455,30 @@ public boolean equals(@Nullable Object object) {
   public static <A, B> Converter<A, B> from(
-      Function<? super A, ? extends B> forwardFunction,
-      Function<? super B, ? extends A> backwardFunction) {
+      Function<@NonNull ? super @NonNull A, @NonNull ? extends @NonNull B> forwardFunction,
```
You don't need to supply both a lower and an upper bound of `@NonNull`; just an upper bound is enough.
(Or did you write them both redundantly on purpose?)

mernst commented on guava/src/com/google/common/base/Enums.java:
```diff
@@ -74,6 +76,9 @@ public static Field getField(Enum<?> enumValue) {
 
   @GwtIncompatible // java.lang.ref.WeakReference
+  @SuppressWarnings({"argument.type.incompatible", "iterating.over.nullable"})
+  // Missing annotated version of class java.util.EnumSet in annotated JDK for nullness
```
Could you make a pull request that adds the missing annotations?

dilraj45 commented on guava/src/com/google/common/base/AbstractIterator.java:
```diff
@@ -73,6 +73,13 @@ private boolean tryToComputeNext() {
 
   @Override
+  @SuppressWarnings("return.type.incompatible") // warning is being issued for difference in super
+  // bound for next and the expected return type. If type is qualified with @Nullable annotation
```
If hasNext() returns true, next will be non-null in case T is declared to be non-null type else it may or may not be null. And so will be the return type. Can this behaviour be expressed using annotations?

dilraj45 replied:
I could only find the ways to confirm `@NonNull` type. I'm not sure how we can do this for ensuring type `T`. Can you please explain how it can be done?
Thanks!

dilraj45 commented on guava/src/com/google/common/base/Absent.java:
```diff
@@ -55,7 +58,9 @@ public T or(T defaultValue) {
 
   @Override
-  public T or(Supplier<? extends T> supplier) {
+  @SuppressWarnings("nullness") //throws NullPointerException in case supplier returns null, which
```
As per the error message, which says the following
>"use Optional.orNull() instead of a Supplier that returns null"

Is it safe to assume that supplier.get will always return a non-null value in this case?

dilraj45 replied:
Are there any annotations similar to RequiresNonNull that can be used here to ensures that this method is being called only when suppplier.get is guaranteed to return a non-null value. What another possible approach can we employ to address this exception? Can it be assumed that 'or' method will only be called for suppliers that return non-null value?

dilraj45 replied:
Only another way I could of think of is to restrict the return type to `@NonNull` for "supplier.get()" and we should not be doing that for, the return type should depend on the type `T`. Further using `@RequiresNonNull` will lead to `contracts.precondition.override.invalid` for overridden class Optional requires no such precondition for method "or". I guess we'll need to suppress the warning. Please correct me if I'm wrong

dilraj45 commented:
@mernst I have addressed your comments. Closing this pull request in reference to other smaller pull requests for the same changes. Please refer to https://github.com/typetools/guava/pull/6, https://github.com/typetools/guava/pull/7, https://github.com/typetools/guava/pull/8, https://github.com/typetools/guava/pull/9 and https://github.com/typetools/guava/pull/11
